lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts

```diff
@@ -73,19 +80,58 @@ export class ServerTelemetryManager {
         return lastRequests.cacheHits;
     }
 
-    getLastRequests(): ServerTelemetryCacheValue { 
-        const initialValue: ServerTelemetryCacheValue = {
-            failedRequests: [],
-            errors: [],
-            errorCount: 0,
-            cacheHits: 0            
-        };
-        const lastRequests = this.cacheManager.getItem(this.telemetryCacheKey, CacheSchemaType.TELEMETRY) as ServerTelemetryCacheValue;
+    /**
+     * Get the server telemetry entity from cache or initialize a new one
+     */
+    getLastRequests(): ServerTelemetryEntity { 
+        const initialValue: ServerTelemetryEntity = ServerTelemetryEntity.initializeServerTelemetryEntity();
+        const lastRequests = this.cacheManager.getItem(this.telemetryCacheKey, CacheSchemaType.TELEMETRY) as ServerTelemetryEntity;
         
         return lastRequests || initialValue;
     }
 
+    /**
+     * Remove server telemetry cache entry
+     */
     clearTelemetryCache(): void {
-        this.cacheManager.removeItem(this.telemetryCacheKey);
+        const lastRequests = this.getLastRequests();
+        const numErrorsFlushed = ServerTelemetryManager.maxErrorsToSend(lastRequests);
+        const errorCount = lastRequests.errors.length;
+        if (numErrorsFlushed === errorCount) {
+            // All errors were sent on last request, clear Telemetry cache
+            this.cacheManager.removeItem(this.telemetryCacheKey);
+        } else {
+            // Partial data was flushed to server, construct a new telemetry cache item with errors that were not flushed
+            const serverTelemEntity = ServerTelemetryEntity.initializeServerTelemetryEntity();
+            serverTelemEntity.failedRequests = lastRequests.failedRequests.slice(numErrorsFlushed*2);
+            serverTelemEntity.errors = lastRequests.errors.slice(numErrorsFlushed);
+            
+            this.cacheManager.setItem(this.telemetryCacheKey, serverTelemEntity, CacheSchemaType.TELEMETRY);
+        }
+    }
+
+    /**
+     * Returns the maximum number of errors that can be flushed to the server in the next network request
+     * @param serverTelemetryEntity 
+     */
+    static maxErrorsToSend(serverTelemetryEntity: ServerTelemetryEntity): number {
+        let i;
+        let maxErrors = 0;
+        let dataSize = 0;
+        const errorCount = serverTelemetryEntity.errors.length;
+        for (i = 0; i < errorCount; i++) {
+            const apiId = serverTelemetryEntity.failedRequests[2*i];
+            const correlationId = serverTelemetryEntity.failedRequests[2*i + 1];
+            const errorCode = serverTelemetryEntity.errors[i];
+            dataSize += apiId.toString().length + correlationId.toString().length + errorCode.length + 3; // Add 3 to account for commas
```

Comment:
Some more comments might be helpful here, not clear to me what/why is happening.

Reply:
Added comments. This function is determining how many of our cached errors can be added to the header sent to the server, based on the number of characters.